File: 26-Dynamic-Programming-Part-II/python/03_subset_sum_equals_target.py
from typing import List

class Solution:
    def isSubsetSum(self, arr: List[int], target: int) -> bool:
        n = len(arr)
        dp = [[False] * (target + 1) for _ in range(n + 1)]

        for i in range(n + 1):
            dp[i][0] = True

        for i in range(1, n + 1):
            for sum_val in range(1, target + 1):
                not_take = dp[i - 1][sum_val]
                take = False
                if arr[i - 1] <= sum_val:
                    take = dp[i - 1][sum_val - arr[i - 1]]

                dp[i][sum_val] = take or not_take

        return dp[n][target]

File: 26-Dynamic-Programming-Part-II/cpp/03-subset-sum-equals-target.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool isSubsetSum(vector<int>& arr, int target) {
        int n = arr.size();
        vector<vector<bool>> dp(n + 1, vector<bool>(target + 1, false));

        for (int i = 0; i <= n; i++) {
            dp[i][0] = true;
        }

        for (int i = 1; i <= n; i++) {
            for (int sum = 1; sum <= target; sum++) {
                bool notTake = dp[i - 1][sum];
                bool take = false;

                if (arr[i - 1] <= sum) {
                    take = dp[i - 1][sum - arr[i - 1]];
                }

                dp[i][sum] = take || notTake;
            }
        }

        return dp[n][target];
    }
};
